tinygrad/tensor.py

```diff
@@ -215,37 +214,42 @@ def slice(self, arg:Sequence[Optional[Tuple[int, int]]]) -> Tensor:
     padding = tuple((max(0, -p[0]), max(0, p[1]-self.shape[i])) for i,p in enumerate(arg_))
     return self.pad(padding).shrink(tuple((p[0] + padding[i][0], p[1] + padding[i][0]) for i,p in enumerate(arg_)))
 
-  # Tensors mostly follow the normal python indexing / slicing behavior for sequences
```

Comment:
Are these comments no longer true? I see you rewrote this function, are all the behaviors tested?

Reply:
They are true (I forgot to include a simpler version of the comments):
- Strides other than 1 raise an `AssertionError`.
- Empty slices are partially supported: forward pass works, but backward for tensors with shape containing zeros in needs to be revised, in order to make it work (idk who needs `.grad` of a zero-in-shape tensor).

The slicing behavior is preserved and tests too + I added some more tests from https://github.com/geohot/tinygrad/pull/506#discussion_r1093620465 and for zero-in-shape tensors.

Reply:
Hmm, shape shouldn't ever actually contain 0s, but now you can have a 0 length shape.

Can you add a (failing) test for this unsupported behavior? It's fine if it's skipped.

Also, the strides should be easy to add support for. Pad+Reshape+Slice+Reshape

Reply:
> Hmm, shape shouldn't ever actually contain 0s, but now you can have a 0 length shape. Is that what doesn't work backward?

Not sure if I understood your question, but backward doesn't work, regardless of 0dim tensors support or anything in this PR. 

I wanted to add support (maybe a separate PR) for a behavior with this kind of tensors similar to the one in PyTorch (e.g. if `pt` has shape, say, `[0, 3, 3]`, then `pt + 1` has also shape `[0, 3, 3]` and also allows to backpropagate towards leaf tensors).

On the other hand, who'd purposefully do bad slices on tensors, followed by some ops followed by backward? Maybe we should settle with just raising an exception in those cases for the moment.

> Can you add a (failing) test for this unsupported behavior? It's fine if it's skipped.

I'll add a commit with such a test.

> Also, the strides should be easy to add support for. Pad+Reshape+Slice+Reshape

I'll have a look at it.